clap_complete/src/dynamic/completer.rs

```diff
@@ -76,6 +77,15 @@ pub fn complete(
         } else if arg.is_escape() {
             is_escaped = true;
             state = ParseState::ValueDone;
+        } else if allow_hyphen {
+            match state {
+                ParseState::ValueDone | ParseState::Pos(..) => {
+                    (state, pos_index) =
+                        parse_positional(current_cmd, pos_index, is_escaped, state);
+                }
+                ParseState::Opt((ref opt, count)) => state = parse_opt(opt, count),
+            };
+            allow_hyphen = false;
         } else if let Some((flag, value)) = arg.to_long() {
```

Comment:
Clap's behavior is a little more subtle than this

You look to be covering the case where `ParseState` is `Opt` or `Pos` (next argument *is* a value).  Another case we need to test for and handle is when we are in the `ValueDone` state, if the long or short is unknown and the next positional allows hyphen values, then we will take the unknown argument and treat it as a value.